### PR TITLE
ICU-20923 Fix compact notation with percent.

### PR DESCRIPTION
**Pull request by sffc**
This PR depends on #1230.  I broke tradition and pushed my branches to upstream instead of to my fork so that I can actually make the PRs depend on each other in GitHub's UI ([details](https://github.com/isaacs/github/issues/959#issuecomment-642838613)).

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20923
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

**Comment by sffc** [comment]
@echeran ping

File: icu4c/source/i18n/number_formatimpl.cpp [hunk 1]
```diff
@@ -144,8 +145,20 @@ NumberFormatterImpl::macrosToMicroGenerator(const MacroProps& macros, bool safe,
...
+    // Use CLDR unit data for all MeasureUnits (not currency and not
+    // no-unit), except use the dedicated percent pattern for percent and
+    // permille. However, use the CLDR unit data for percent/permille if a
+    // long name was requested OR if compact notation is being used, since
+    // compact notation overrides the middle modifier (micros.modMiddle)
+    // normally used for the percent pattern.
+    bool isCldrUnit = !isCurrency
+        && !isBaseUnit
+        && (unitWidth == UNUM_UNIT_WIDTH_FULL_NAME
+            || (
```
Comment (echeran) [inline icu4c/source/i18n/number_formatimpl.cpp:157]: optional:  rewrite
```
...
|| (
  !(isPercent || isPermille)
  || isCompactNotation
)
...
```
as
```
...
|| !(isPercent || isPermille)
|| isCompactNotation
...
```

**Comment by sffc** [comment]
^ I rebased the branch.  I will rebase one more time after #1230 is in.

File: icu4c/source/i18n/number_formatimpl.cpp (hunk 1 above)
Comment (sffc) [reply icu4c/source/i18n/number_formatimpl.cpp]: Done.

**Pull request merged**